### DISCUSSION TITLE
Hand siteswap, MHN, and polyrhythm patterns

### DISCUSSION THREAD

**Discussion by jkboyce** [jkboyce/jugglinglab#91] [General]
(Migrated from discussion on issue #87 which is now unrelated)

@froudaut and @JigyasuJuggler : very interesting discussion!

1. I agree with @JigyasuJuggler 's remark that MHN notation (and juggling matrices as Polster describes on pages 85-90) provide a complete description here. In the case of unequal beat times for the jugglers/hands, just choose a shorter beat that divides evenly into every hand's beat period.

2. Caveat 1: The above assumes that all throw times are rational numbers. One could break this model with for example juggler 1 throwing on beats 1, 2, 3, ... and juggler 2 throwing on beats 1+(1/pi), 2+(1/pi), ... That said, an irrational throwing schedule like this won't change the patterns you enumerate, only the heights of throws. So in that sense there isn't anything to be gained by including non-rational throw times.

3. Caveat 2: Polster in his juggling matrix discussion assumes unlabelled objects. With labelled objects we get the question of object assignment in multiplexed throws. Thank you @froudaut for linking Miika Toukola's paper. The multiplexing rules proposed there aren't fully complete (there are some tiebreaker corner cases he doesn't discuss) but overall they are similar to the rules used by Juggling Lab. The relevant code is [here](https://github.com/jkboyce/jugglinglab/blob/master/source/jugglinglab/notation/MHNPattern.java#L611).

4. Writing down matrices for everything is cumbersome. So the question arises, how to describe these matrices in human-readable form? In Juggling Lab I've tried to adhere to "simple things should be simple, complex things should be possible." I'm not 100% sure I achieved that, especially with mixed sync/async patterns like:
`55556x56x1x(6x,4)(4,6x)(6x,4)(5x,7)(6x,4)`
which are difficult to write down and debug. In such cases it might be easier to write:
`(0,5)!(5,0)!(0,5)!(5,0)!(0,6x)!(5,0)!(0,6x)!(1x,0)!(6x,4)(4,6x)(6x,4)(5x,7)(6x,4)`

5. An idea I thought about for Juggling Lab, but haven't implemented, is to make the juggling matrix visible to the user somehow. I find that when I'm writing complicated patterns, I always pull out a pencil and paper and start drawing arrows on a grid, and only when I have that working do I translate it into a pattern to type in. It might be interesting to have some way of doing that directly in the program, although I don't know what form that would take.
One possibility would be a view where the user types in a pattern, and as they type a matrix view below redraws in real time and highlights any conflicts. The user can edit in-place and try to get the errors to go away. Potentially it could also work the other direction: You edit the matrix with your mouse and it rewrites the siteswap description.

6. I like HSS because it takes a large class of patterns that people care about--especially passing patterns--and makes them easy and concise to describe.

7. Technical point, @froudaut I saw on your polyrhythms page (excellent page by the way) "3x Hold Times are not handled by JugglingLab". In JL any throw to the same hand can be held by putting an H after it. So you can write for example:
`(4,7x)(4,3xH)(4,0)!(0,7x)!(4,0)!(0,3xH)!(4,0)`

8. Technical point 2, I'm adding a change to the JL renderer that will make these types of short-beat MHN patterns animate more smoothly. The current method puts too many empty events into the pattern JML, which makes the hand movement very "robotic". It's especially noticeable in some of the hss examples.

9. Technical point 3, @JigyasuJuggler the [`splitcatchfactor`](https://github.com/jkboyce/jugglinglab/blob/master/source/jugglinglab/notation/MHNPattern.java#L788) that you mention can be set to 0 if you want all catches on a beat to land at the same time. I will expose this as a user setting so that squeeze patterns like `([42],4x)*` can be animated correctly. Anyone want to propose a name for this parameter? squeezebeats? squeezecatchbeats? :)

**Comment by froudaut** [comment]
Hello,


thanks again both of you for all your clarification.
And sorry for my english, I'm french ;-/

If I may reply something concerning your email :

@JigyasuJuggler, @jboyce : for sure MHN is the killer notation in such context. However most of the time I prefer classical SS for solo 
juggling (sync, async, mult ...) : it is easier to share, to understand and most of the time only the concerned beats are described whereas in 
MHN all the beats have to be written. And indeed it will be great to have a way to enter and adapt MHN matrix in jugglingLab.

@JigyasuJuggler : concerning the rythm you describe with a not fixed tempo I do not believe you have to think in that way. The polyrythm 
terminology comes from music so I think we do not modify this terminology. Here is a basic explanation for such rythm in percussion: 
https://www.jonglage.net/theorie/notation/siteswap-avancee/refs/Brad%20Meyer-Polyrhythm-FUNdamental.pdf
A more general description for music may be found here : 
https://escholarship.org/content/qt8m46811d/qt8m46811d_noSplash_5ec7a12d1de32a645a095a03e6bdfd0b.pdf?t=oeitm5

In music a partition is split in measures (or bars) and time signatures describe for each bar the number of beats. I guess that for juggling it 
has to be the same and each tempo has to be seen in an atomic way.

@jboyce :
- you ask for the mixed notation implementation status in jugglinglab : indeed I tried a lot of things and it works pretty well. Just a few 
things are not working. I should check but I think that using * in the middle of the pattern or ! for vanilla siteswap gives errors.
- Thanks for pointing my mistake on holding object with a throw greater than 2 . I hadn’t seen in the doc, but indeed it is described.
- for colorization in multiplex, indeed this quite similar to what jugglinglab does with a few differences: I thus guess that color 
ordering in the multiplex is done according to the previous throw in the same ordering (not checked on recent version however).  There were then a few issues :

        - it was dependent from the circular permutation.
         Ex : 32322 and 22[32]3  
![image](https://user-images.githubusercontent.com/26428344/148619210-57c93c62-aea0-4ba5-a12c-f57f864f43ce.png)
![image](https://user-images.githubusercontent.com/26428344/148619251-f06b2a28-0232-4fed-8ef4-bab52a33ad0b.png)

-	When wildcard * is used, it is not clear how it behaves after the ss period :
       Ex with hurry on async : [23]* <=> [23][32] or [23][23] ….
       For example I notice that the behavior was :
            [23]* <=> [23][32]…
       And it was strangely also the same for :
            [32]* <=> [23][32]…

       The same is true for sync with multiplex. How it behaves when expanding the ss after the * ?
           Ex: ([22x],2)* <=> ([22x],2)([2x2],2) or  ([22x],2)([22x],2) or whatever other possibilities.
       As with hurry on async, there is not a single rule and according to the heights and the ordering you choose in the multiplex, the behavior was different on JL

The definition of Miika Toukola avoids some of these issues. We have however to define how to expand ss after the * and the behavior if two previous throws that leads to the multiplex have same height. Another thing to add is a ordering between a throw and the same with an ‘X’


And lastly, because you wrote on new features on JL such as MHN, another one could be to a way to add or remove an object during the pattern. I heard about two ways to do that :

- one is what is called antisiteswap or originally “High energy Juggling” : 
https://www.jonglage.net/theorie/notation/siteswap-avance/refs/Andrew%20Conway%20-%20High%20Energy%20Juggling%20(1994)%20-%20rec.juggling.pdf
A tool exists on windows to emulate this : http://vcg.isti.cnr.it/~tarini/index.php?3?juggling.html
In video it gives : http://juggling.tv/1800
However although it is written in the video that it has mathematically no impact, it is not true. The average theorem does not apply most of 
the time.

- the other solution is based on a note made by Charlie Dancey : 
https://www.jonglage.net/theorie/notation/ladder/refs/Charlie%20Dancey%20-%20Siteswaps,%20State%20Diagrams,%20and%20New%20Ideas.pdf
In his proposal, state-transition matrix have not to be limited to a given objects number. Adding symbol such as – or + in this matrix may 
symbolize it.
Again with this model the average theorem does not apply anymore.

--
Fred

Le 2022-01-07 01:27, Jack Boyce a écrit :

**Reply by JigyasuJuggler** [reply]
@froudaut: I think I did not explain myself properly. I was not suggesting changing the terminology. What I was trying to say with the rhythm I described was that no matter how many categorizations or terminologies we employ, someone is sooner or later likely to invent a pattern that doesn't fit into any of the known categories (like the rhythm I gave. Or as far as I can make out, even Richard Kohut's pattern in Nickmans' article, as mentioned in part 2 of my link). So inventing new terms for every new type of juggling rhythm has its limitations.

Thanks for sharing some more interesting and informative links! On drops and pick-ups, one way to think of this _in theory_ (while staying consistent with average and permutation tests) could be to designate an additional 'hand' where the dropped ball lands and is eventually picked up from. If this drop and pick-up is repetitive, then again MHN or oss/hss (for async. patterns) should be able to handle it.

**Reply by jkboyce** [reply]
@froudaut Apologies for the slow response. Thank you for investigating orbits in multiplex patterns. I've logged two issues (#110 and #111) to track the problems you noted. At some point I intend to clean up (or at least document) how Juggling Lab handles these situations.

@JigyasuJuggler Regarding your proposal for pickups/drops, an invisible juggler can sort of emulate this:
`pattern=<333p 330330|00000204p 0>;hidejugglers=2`. The notation for this, and how it's rendered, could obviously be improved if we want to add pickups/drops/stalls to Juggling Lab more comprehensively.

**Comment by froudaut** [comment]
Hi,

 

After some more thinking on OSS/HSS/Polyrhythms ... I have a few comments.

-  First is concerning Galoped Polyrhythms, Galoped Async and Multi-Tempo Async (i.e the throws are async but the tempo is not fixed on each hand and between hands. This is quite similar to Galoped Polyrythms but the ratio is hard to get) :
@jigyasujuggler in your blog:
https://jigyasujuggler.com/2019/03/25/vanilla-siteswap-extensions-an-alternative-approach-i/ you argue more and less that OSS/HSS is the only way to describe the example you give. I call it Multi-Tempo Asynchronous
If I draw a ladder diagram of such example after finding a common Tempo for hands and throws (as we basically do in such case) it gives: 
![(a,0)(0,0)(0,0)!(9,0)!(0,9)(0,0)(a,0)(0,0)(0,7x)!](https://user-images.githubusercontent.com/26428344/153570257-601d2ee7-ea6d-40db-b5f6-a6b180d0ef3f.png)

It is then possible to simulate this in JugglingLab using Mixed Siteswap : (a,0)(0,0)(0,0)!(9,0)!(0,9)(0,0)(a,0)(0,0)(0,7x)!
Then with 10bps we exactly have the same animation you provide :
![(a,0)(0,0)(0,0)!(9,0)!(0,9)(0,0)(a,0)(0,0)(0,7x)!](https://user-images.githubusercontent.com/26428344/153570335-a86870ee-fbf0-4963-8a68-0a3c57b83426.gif)

Using OSS/HSS finally you have :
OSS: a000099000a0007
HSS: 500005800050007
And you have to define the apparition order of Hands : ie R, L

Thus with this example I am not really agree that OSS/HSS is easier to understand and more concise. Most of the time the fictive beats you add in Mixed notation will be translated into HSS and OSS for Galoped Async/Multi-Tempo Async and Galoped Polyrythms [I have done the exercice on all the examples you provide in the blog, thanks for that ;-)] With more than 2 hands, it will also be the case in comparison to MHN. But you are right, we avoid the use of "X" that are quite confusing and we are then able to draw Siteswap Async diagram.  
![hss_a000099000a0007_hss500005800050007](https://user-images.githubusercontent.com/26428344/153684889-53fcab54-7002-4be3-83a4-44eadf6d6369.png)


I then think that OSS/HSS is really great for patterns quite basic with more than 2 hands, to describe and also to discover them. Thus I especially think about what is described into Polster’s book concerning Multihand Cascades and Fountains.

It is then really easy to understand that a 5 balls cascade with 4 hands may be described by OSS=5/HSS=4 with the corresponding hands order. Some patterns with some more sophisticated HSS may also be very convenient to describe with such notation.
 
![multihand-cascades-fontaines-map](https://user-images.githubusercontent.com/26428344/153571377-7bfe7b76-059d-4113-892e-9380c99ee9e9.png)


- Again, but since I have not written it on Github ... this is concerning the multiplexing in OSS. In the Shannon demonstration and its extension to non-uniform patterns, there is no mention of multiplexing. It however seems that there is a nice duality in such cases between objects and hands and we then may extend OSS with multiplexes. I do not think that an extra mechanism to attach hands to orbits in a multiplex throw, other than what is already done has to be added.   
![hss_24 54 _hss31](https://user-images.githubusercontent.com/26428344/153571412-47163a9b-42f1-4105-80dc-5a2dcf55a537.png)


- A thing I also noticed in the current implementation is the fact that now some throws higher that 2 are held into hand.
If you try for example: OSS=504/HSS=501 you will see that 4 are no more throwed
Or try with OSS=7531 and HSS=31, you will notice that 3 are no more throwed
![oss 504 hss 501](https://user-images.githubusercontent.com/26428344/153573515-16e55f13-21e5-469c-aa5e-d95d16e914e8.gif)


-   Also sometimes heights difference are no more reflected. Check OSS=7531 and HSS=31. You will see that 7 and 5 have the same height. The Dwell Time does not seem to stay uniform to permit this.  
![hss_7531_hss31](https://user-images.githubusercontent.com/26428344/153571679-80fdaa56-fb43-406c-a956-6cca1c92b0bf.gif)

 
-  Since I wrote about adding Multiplexes in OSS, we also may think on Multiplexes in HSS. In that case it means that throws are done at the same time by multiple hands. It does not fit with the HSS/OSS Model but it could fit with some specific MHN matrix on which rows are not mapped to a specific hand. There is perhaps something to think about to have a similar model to simplify MHN matrix. I will try to think about this.

**Comment by JigyasuJuggler** [comment]
@froudaut : Thanks for your detailed study and analysis and for bringing up several interesting points as usual :).

1. It was not my intention to say that OSS/HSS is the **_only_** way to describe the given example. In fact, as @jkboyce had mentioned initially, the implementation of the OSS/HSS feature in Juggling Lab internally converts the OSS/HSS I derived in the blog to the synchronous pattern you derived in your comment and then does the simulation (which explains why you got the exact same result). What I am trying to argue in the blog is that the OSS/HSS representation allows us to live with the vanilla async notation for ALL async patterns. Not only that, we need not mess up "everything that was a clear and logical theory" as happens in the JoePass notation. The clear and logical theory remains clear and logical. This also makes it easier to check for pattern validity by using the standard average and permutation tests and as you said, it is also easier to invent new patterns. This is not to say that the other terminology or representations should be abandoned completely. For example, use of hand specifiers, synchronous notation, "x", * (for hurries as in JoePass) etc. make it quite easy to visualize how a pattern will look like just by looking at the notation. Similarly, the terminology like galloped or polyrhythm gives a sense of what the pattern will feel like. On the other hand, a pattern in the OSS/HSS notation is usually not easy to visualize. But then that's where the simulator helps - we can quickly invent new patterns by combining any two vanilla siteswaps (oss and hss) and simulate it to see what it looks like. 

2. If we are willing to let the ambiguity in the identification of the multiplexed hands be resolved as it is done for multiplexed objects in JL, then yes, in theory, I think multiplexed hands could also be introduced. In the OSS/HSS case as I said, I simply covert the OSS/HSS to existing sync notation supported in JL to simulate it. I could not do that for multiplexed hands.

3. In the usual siteswap notation, the "2" throws are a hold. You will also notice that a "1x" throw is a hold. These are in fact special cases of the general rule that whenever the OSS and HSS number on a beat are the same, the throw can be held. The "usual" HSS notation is basically hss = 2, so whenever the OSS is 2, the throw is held. Similarly for a 1x throw, it means the object has to be thrown again after one beat by the same hand. So the hand might as well hold the object. The idea is that if on a particular beat, the OSS is N and the HSS is also N, then it means that the object is not getting thrown again by any other hand for another N beats and also that the throwing hand is not throwing anything else for another N beats. So the hand can afford to just hold the object. In the OSS=504/HSS=501 case, it is actually the "5" that is not getting thrown, the 4 is a crossing throw because of the hand sequencing and is the one getting thrown. In the OSS=7531/HSS=31 case, both the 3 and the 1 are not getting thrown because the hss and oss siteswap numbers match on those beats. If you change to OSS=7531/HSS=13, you will find that now the 3's and the 1's are thrown because now when the oss is 3, the hss is 1 and vice-versa so the condition for a "hold" is no longer met.

4. The reason for the behaviour of siteswap throws not going to the "expected" height is exactly as you guessed - the dwell time is not constant. This behaviour can be turned off by doing the manual setting: dwellmax = false. Actually, though I spoke of "hold" in point 3 above, it is the default dwellmax = true setting which is causing an _apparent_ hold. For example, in the OSS=504/HSS=501 if you slow down the bps to 2 and look carefully, you can see a small toss happening for the 5. When hss = 2 (the "usual" case), the range of available dwell times is 0 to 2 beats and JL default behaviour is to set it to 1.3 beats (except for 1 throws), i.e., the hand remains empty for 0.7 beats after which it catches the object it has to throw next. In the OSS/HSS scheme, if a hand is throwing next after N beats, then it has a potential dwell beats range from 0 to N beats - the next object can be caught at any time in this range. The OSS/HSS code basically again allows the hand to remain empty for 0.7 beats (which is why there is a "small toss") after which it catches the next object to be thrown making the dwell beats for that object to be N-0.7 beats (Note: In the "usual" case, N = 2 and dwell beats = 2 - 0.7 = 1.3 beats). As I said earlier, you can turn off this feature by setting the manual settings for your particular example as: hss = 31; dwellmax = false. There is a further option where you can set the dwellmax to false but retain the option of "holding" objects for which oss = hss on a certain beat (something like the "H" modifier that @jkboyce had mentioned to allow holding a 3x by using 3xH somewhere in this thread itself if I remember correctly). So for example, you could try oss = 7531 with the manual settings being hss = 31; dwellmax = false; hold = true. This will make the 7 go higher than the 5 (because the dwellmax has been made false) while still allowing the 3 and 1 to be held (because hold is true). Note that the hold=true makes it a genuine hold unlike the dwellmax induced hold which actually has a small toss. Interestingly, if you try OSS=7531/HSS=13, the 5 and the 3 are now thrown to the same height.

The HSS documentation (https://jugglinglab.org/html/HandSiteswapFeature.pdf) does talk about the dwellmax and hold parameters briefly on slide 7. We are in the process of making this documentation more detailed though our progress is very slow :(

5. Hand multiplexing - As mentioned in my response number 2 above, I agree that it could be done in theory. To be a perfect dual of object multiplexing though, I would think that hand multiplexing would mean that more than one hands throw the _same_ object on the same beat. Indeed, objects and hands could both be simultaneously multiplexed but this could quickly lead to confusion whether some objects are being thrown as multiplex or as synchronous throw! I had briefly toyed with this idea in another one of my blogs (https://jigyasujuggler.com/2018/08/31/siteswap-notation-iv-multiplex-patterns/, see the section titled "Interesting throws") but had dismissed hand multiplexing as an "uninteresting" idea then. Maybe it is time to revisit that conclusion. Some examples of what I think will qualify as throws with multiplexed hands are also shown in the first video embedded in this blog @ 1:30 to 1:45 and then again @2:34 to 3:24 (see https://youtu.be/31BblUAE8ws). 

I hope points 3 and 4 above have helped clarify your observations on the "unexpected" holds and strange throw heights that you observed when playing with the OSS/HSS feature. If not, then I think if you play around with dwellmax = true/false and hold = true/false options, you'll be able to figure out what is happening.

**Reply by froudaut** [reply]
@JigyasuJuggler . Ok thanks a lot for your clarification . I got it I think ;-)
